src/bls12_381/ec.rs

```diff
@@ -2012,6 +2062,33 @@ pub mod g2 {
         );
     }
 
+    #[test]
+    fn test_g2_scale_by_cofactor() {
+        use rand::{Rng, SeedableRng, XorShiftRng};
+
+        fn random_point<R: Rng>(rng: &mut R) -> G2Affine {
+            loop {
+                let x = rng.gen();
+                let greatest = rng.gen();
+
+                if let Some(p) = G2Affine::get_point_from_x(x, greatest) {
+                    if !p.is_zero() {
+                        return p;
+                    }
+                }
+            }
+        }
+
+        let mut rng = XorShiftRng::from_seed([0x5dbe6259, 0x8d313d76, 0x3237db17, 0xe5bc0654]);
+        for _i in 0 .. 20 {
+            let p = random_point(&mut rng);
+            let got = p.scale_by_cofactor().into_affine();
+            assert!(got.is_on_curve());
+            assert!(got.is_in_correct_subgroup_assuming_on_curve());
+        }
+    }
```

Comment:
I think we should also test that the new `scale_by_cofactor` matches the old one for random curve points (not necessarily in the subgroup).